storage/innobase/fil/fil0crypt.cc

```diff
@@ -467,16 +470,88 @@ fil_space_crypt_t *fil_space_read_crypt_data(const page_size_t &page_size,
   crypt_data->type = type;
   crypt_data->min_key_version = min_key_version;
   crypt_data->encrypting_with_key_version = min_key_version;
-  crypt_data->page0_offset = offset;
-  memcpy(crypt_data->iv, page + offset + bytes_read, iv_length);
+  crypt_data->private_version = 1;
+  memcpy(crypt_data->iv, page + offset + bytes_read, CRYPT_SCHEME_1_IV_LEN);
+  bytes_read += CRYPT_SCHEME_1_IV_LEN;
 
-  bytes_read += iv_length;
+  crypt_data->encryption_rotation = static_cast<Encryption_rotation>(
+      mach_read_from_4(page + offset + bytes_read));
+  bytes_read += 4;
 
-  crypt_data->encryption_rotation =
-      (Encryption::Encryption_rotation)mach_read_from_4(page + offset +
-                                                        bytes_read);
+  uchar tablespace_key[ENCRYPTION_KEY_LEN];
+  memcpy(tablespace_key, page + offset + bytes_read, ENCRYPTION_KEY_LEN);
+  bytes_read += ENCRYPTION_KEY_LEN;
+
+  if (std::search_n(tablespace_key, tablespace_key + ENCRYPTION_KEY_LEN,
+                    ENCRYPTION_KEY_LEN,
+                    0) ==
+      tablespace_key) {  // tablespace_key is all zeroes which means there is no
+                         // tablepsace in mtr log
+    crypt_data->set_tablespace_key(nullptr);
+  } else {
+    crypt_data->set_tablespace_key(
+        tablespace_key);  // We are using the same iv for both
+                          // MK encryption and KEYRING encryption
+  }
+
+  return crypt_data;
+}
+
+static fil_space_crypt_t *fil_space_read_crypt_data_v2(
+    const page_size_t &page_size, const byte *page) {
+  const ulint offset = fsp_header_get_keyring_encryption_offset(page_size);
+
+  if (memcmp(page + offset, ENCRYPTION_KEY_MAGIC_PS_V2,
+             ENCRYPTION_MAGIC_SIZE) != 0) {
+    /* Crypt data is not stored. */
+    return nullptr;
+  }
+
+  ulint bytes_read{ENCRYPTION_MAGIC_SIZE};
+
+  uint8_t type = mach_read_from_1(page + offset + bytes_read);
+  bytes_read += 1;
```

Comment:
Can you please add this assert (like you add in parse)


  ut_a(type == CRYPT_SCHEME_UNENCRYPTED ||
       type == CRYPT_SCHEME_1);  // only supported

Reply:
done